File: JAVA/Leetcode/ParitySwap.java
/*
Problem: 2840. Check if Strings Can be Made Equal With Operations II

--------------------------------------------------

🧠 UNDERSTANDING THE OPERATION:

We can swap indices i and j such that:
(j - i) is EVEN

👉 This means:
Both indices must have SAME PARITY

So:
- Even index ↔ Even index
- Odd index ↔ Odd index

--------------------------------------------------

💡 CORE INSIGHT:

We can rearrange:
✔ All EVEN positions among themselves
✔ All ODD positions among themselves

BUT:
❌ Cannot mix even ↔ odd

--------------------------------------------------

🎯 GOAL:

To make s1 == s2,
we must ensure:

1. Even index characters match (frequency-wise)
2. Odd index characters match (frequency-wise)

--------------------------------------------------

🧪 HOW WE CHECK:

We use frequency arrays:

evenCounts[char] → tracks difference between s1 and s2 at even indices
oddCounts[char]  → tracks difference at odd indices

Process:
- Increment for s1
- Decrement for s2

--------------------------------------------------

✅ If all counts become 0:
→ Perfect match → return true

❌ Else:
→ Impossible → return false

--------------------------------------------------

⏱ Complexity:

Time: O(n)
Space: O(1) (since only 26 letters)

--------------------------------------------------
*/

import java.util.*;

class ParitySwap {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        String s1 = sc.next();
        String s2 = sc.next();

        System.out.println(checkStrings(s1, s2));

        sc.close();
    }

    public static boolean checkStrings(String s1, String s2) {

        int[] evenCounts = new int[26];
        int[] oddCounts = new int[26];

        for (int i = 0; i < s1.length(); i++) {

            // Even index group
            if (i % 2 == 0) {
                evenCounts[s1.charAt(i) - 'a']++;
                evenCounts[s2.charAt(i) - 'a']--;
            }
            // Odd index group
            else {
                oddCounts[s1.charAt(i) - 'a']++;
                oddCounts[s2.charAt(i) - 'a']--;
            }
        }

        // Validate both groups
        for (int i = 0; i < 26; i++) {
            if (evenCounts[i] != 0 || oddCounts[i] != 0) {
                return false;
            }
        }

        return true;
    }
}

File: PYTHON/ParitySwap.py
"""
Problem: Check if Strings Can be Made Equal With Operations II

--------------------------------------------------

🧠 CORE IDEA:

Swaps allowed only when:
(j - i) is EVEN

👉 So indices must have same parity

--------------------------------------------------

💡 KEY OBSERVATION:

We can independently rearrange:
- Even indices
- Odd indices

So this becomes:
👉 Compare frequency of characters in both groups

--------------------------------------------------

🎯 APPROACH:

Use two frequency arrays:
- even_counts → for index 0,2,4,...
- odd_counts  → for index 1,3,5,...

For each index:
- Add s1 char
- Subtract s2 char

--------------------------------------------------

✅ If all values = 0 → possible
❌ Else → not possible

--------------------------------------------------
"""

def check_strings(s1, s2):

    even_counts = [0] * 26
    odd_counts = [0] * 26

    for i in range(len(s1)):

        if i % 2 == 0:
            even_counts[ord(s1[i]) - ord('a')] += 1
            even_counts[ord(s2[i]) - ord('a')] -= 1
        else:
            odd_counts[ord(s1[i]) - ord('a')] += 1
            odd_counts[ord(s2[i]) - ord('a')] -= 1

    # Validate
    for i in range(26):
        if even_counts[i] != 0 or odd_counts[i] != 0:
            return False

    return True


# ----------- User Input -----------

s1 = input().strip()
s2 = input().strip()

print(check_strings(s1, s2))